src/rpc/rpc_client_utils.js

```diff
@@ -0,0 +1,104 @@
+/* Copyright (C) 2016 NooBaa */
+'use strict';
+
+const RPC_BUFFERS = require('./rpc_request').RPC_BUFFERS;
+
+const schema_client_cache = new WeakMap();
+
+function empty_client_factory() {
+    return {};
+}
+
+function client_factory_from_schema(schema) {
+    if (!schema.compiled) {
+        throw new Error('Invalid schema: schema is not compiled');
+    }
+
+     // If a factory can be found in the cache we provide it from there.
+    const client_factory = schema_client_cache.get(schema);
```

Comment:
Is client_factory_from_schema getting called more than once per schema?
How about moving to RpcSchema to "compile a client"?

Reply:
@guymguym I started by having this code inside the RpcSchema but it seems like the wrong place to place it as It has nothing to do with the scehma itself so I added a seperate utils file.  
But as you observed there is nothing holding us from putting it in  RpcSchema.

Please note that in that implementation the schema will hold the factory, where here we just weakly key the factory with the schema as key

Reply:
Why do you say that it has nothing to do with the schema itself? 
This entire client_proto is being built from the schema definitions.
So yeah it makes a lot of sense to me as a functionality there.

Reply:
@guymguym I changed the code to provide clients directly from the schema object via a `new_client(rpc, options)` method. This makes the client factory creation an optimization/implementation detail of the RPCSchema class and simplify even more the code inside the RPC's new_client code.

I hope you will find this change to your liking